torch/nested/_internal/ops.py

```diff
@@ -847,28 +843,82 @@ def is_same_size_default(func, *args, **kwargs):
+    torch.ops.aten.sum.dim_IntList,
+    "self: jt_all, dim: any?, keepdim: any?, dtype: any?",
 )
 def sum_dim_IntList(func, *args, **kwargs):
-    # sum_dim_IntList can produce a NT or a T depending on whether the ragged dims
-    # are reduced away.
+    """
+    Performs a sum along the provided tensor dimension.
+    Returns a dense tensor if the ragged dimension is reduced away, else returns a nested tensor.
+    """
     _, new_kwargs = normalize_function(
         func, args=args, kwargs=kwargs, normalize_to_only_use_kwargs=True
     )
     inp = new_kwargs.pop("input")
-    assert inp._ragged_idx == 1
-    new_kwargs["dim"], ragged_reduced_away = _wrap_jagged_dims(
-        inp.dim(), new_kwargs["dim"], "sum"
+
+    (
+        new_kwargs["dim"],
+        reduce_on_batch,
+        reduce_on_ragged,
+        reduce_on_non_batch,  # noqa: UFMT
+    ) = _wrap_jagged_dims(
+        inp.dim(),
+        new_kwargs["dim"],
+        "sum",
+        inp._ragged_idx,
     )
 
-    if not ragged_reduced_away:
-        return NestedTensor(func(inp._values, **new_kwargs), **extract_kwargs(inp))
-    else:
-        # Don't wrap because we reduced away the raggedness
-        out = func(inp._values, **new_kwargs)
+    if inp._ragged_idx != 1:
+        raise RuntimeError("sum(): not supported when ragged_idx != 1 for NestedTensor")
+
+    if reduce_on_ragged and inp._lengths is not None:
+        raise RuntimeError(
+            "sum(): not supported for nested tensors with holes, i.e. when lengths is not None, "
+            + "if reducing across the ragged dimension for NestedTensor"
+        )
+
+    if reduce_on_ragged:  # raggedness reduced away --> return dense tensor
+        if (
+            reduce_on_batch
+        ):  # reduction cases: (batch, ragged), (batch, ragged, non-batch), etc.
+            out = func(
+                inp._values, **new_kwargs
+            )  # no need to read offsets --> apply sum directly on values
+        else:
+            if (
+                reduce_on_non_batch
+            ):  # invalid reduction cases: (ragged, non-batch), etc.
+                raise RuntimeError(
+                    "sum(): not supported along a ragged and non-batch dimension for NestedTensor"
+                )
+            # reduction cases: (ragged)
+            out = torch.sum(
+                torch.ops.aten._jagged_to_padded_dense_forward(
+                    inp._values.view(*inp._values.shape[: inp._ragged_idx], -1),
+                    [inp._offsets],
+                    max_lengths=[inp._max_seqlen],
+                ).view(
+                    *inp.shape[: inp._ragged_idx],
+                    inp._max_seqlen,
+                    *inp.shape[inp._ragged_idx + 1 :],
+                ),
+                dim=inp._ragged_idx,
+            )  # need to read offsets --> pad jagged dimension and apply sum
+
         if new_kwargs["keepdim"]:
             out = out.unsqueeze(0)
```

Comment:
note: I don't think this is actually right? e.g. you'll end up with a size of (1, ...) which I don't think matches the semantics of normal keepdim; I think you'd usually expect the output shape to be the same as the input shape (realize this is pre-existing before your PR)

maybe can you open a github issue for this with some details to track this?

Reply:
offline discussion: this is actually fine, doesn't need an issue

Reply:
resolved offline! the pytorch documentation says that intended functionality for `keepdim=True` is for the number of dimensions in the input and output to match, rather than the shapes themselves to match

```
If keepdim is True, the output tensor is of the same size as input except in the dimension(s) dim where it is of size 1.
```